### ISSUE TITLE
avg_pool2d global fast path: padded ROW_MAJOR input is silently wrong for specific shapes

### ISSUE THREAD

**Issue by iwroszTT** [tenstorrent/tt-metal#55227]
Follow-up to #54410, see comment https://github.com/tenstorrent/tt-metal/pull/54410#pullrequestreview-5089018626.

## Summary

#54410 adds a TT_FATAL in reduce_op.cpp that accepts a padded ROW_MAJOR input only in a very specific scenario (i.e., when N and C fold to a single slice and H is padded by no more than the rows needed to reach the next tile boundary).

That closes the reported resnet50 issue, but the same interleaved-pad corruption is still reachable through `avg_pool2d` /` global_avg_pool2d` for a large set of shapes.

 The dense ROW_MAJOR reduce indexes input pages from the tensor's logical row count (not its padded one). In the single-slice case the guard accepts, that means it reads the first H pages and nothing else, which is correct only when every pad row sits after every real row. But a shape records how much padding a tensor has, not where that padding sits. So the guard checks the amount and infers the position: if H is padded by exactly what tile alignment would add, it assumes the pad trails.

Additional debug info (courtesy of Claude):

## Reproduction

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)
for (N, H, W, C) in [(1, 2, 17, 64), (1, 4, 29, 64), (1, 7, 31, 64), (1, 2, 31, 64)]:
    act = torch.ones([N, H, W, C], dtype=torch.bfloat16).float()
    tt = ttnn.Tensor(act, ttnn.bfloat16).pad_to_tile(0.0).to(device)
    got = ttnn.to_torch(ttnn.global_avg_pool2d(tt)).float().reshape(-1)[0].item()
    print(f"NHWC {[N,H,W,C]} padded={list(tt.padded_shape)}  mean(ones) = {got}  (expected 1.0)")
ttnn.close_device(device)
```

## Observed

Legacy `pad_to_tile` ROW_MAJOR NHWC input, `ones` fill, expected mean `1.0`:

| NHWC input | canonical logical / padded | #54410 guard | result |
|---|---|---|---|
| `[1,7,7,64]` — the case #54410 fixes | `[1,1,49,64]` / `[1,1,224,64]` | rejects | 0.285 |
| `[1,2,17,64]` | `[1,1,34,64]` / `[1,1,64,64]` | **passes** | **0.559** |
| `[1,4,29,64]` | `[1,1,116,64]` / `[1,1,128,64]` | **passes** | **0.922** |
| `[1,7,31,64]` | `[1,1,217,64]` / `[1,1,224,64]` | **passes** | **0.973** |
| `[1,2,31,64]` | `[1,1,62,64]` / `[1,1,64,64]` | **passes** | **0.984** |

Worst observed error is 44%. There are 328 such `(H, W)` pairs for `H <= 64, W <= 128`.

## Root cause

`pool2d()`'s global-average fast path folds `(N, H, W)` into a single reduce axis with an explicit logical+padded view (`ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:1110-1121`):

```cpp
uint32_t total_padded_spatial = post_padded[0] * post_padded[1] * post_padded[2];
uint32_t hw_padded_per_batch  = total_padded_spatial / batch_size;
ttnn::Shape canonical_padded({batch_size, 1, hw_padded_per_batch, channels_padded});
Tensor canonical = ttnn::reshape(input, canonical_logical, canonical_padded);
```

`pad_to_tile` pads the last two dims, so an NHWC tensor carries `W_pad` **once per H row** — the pad is interleaved, not a suffix. Before the fold that shows up as `NC = padded[-4] * padded[-3] = N*H != 1`, and #54410's guard rejects it correctly (this is what `test_rm_reduce_padded_h_slices_rejected` pins). The fold sets `NC = batch_size = 1` and collapses all the pad into the H total, leaving only the roundup arm of the predicate:

```cpp
padded[-2] == logical[-2] ||
    (padded[-4] * padded[-3] == 1 && padded[-2] == div_up(logical[-2], tile_h) * tile_h)
```

which passes whenever `H * ceil32(W) == ceil32(H*W)`. The resnet50 shape misses it by a wide margin (49 logical rows in 224 padded), which is why #54410 catches that one; smaller `H`/`W` combinations land on it exactly.

The dense RM reader then reads pages `nc * H_logical + r` for `r in [0, H_logical)` (`reader_unary_reduce_rm.cpp`, H branch), i.e. the first `H*W` pages of the buffer — which straddles the interleaved pad. For `[1,2,17,64]`: pages 0-16 real, 17-31 pad, 32-33 real → 19 real rows summed and divided by 34 → `19/34 = 0.559`.

## Why this cannot be fixed in `reduce_op.cpp`

Two different inputs produce byte-identical canonical tensor specs:

```
NHWC [1,1,62,64] padded=[1,1,64,64] -> canonical [1,1,62,64] / [1,1,64,64] | mean(ones) = 1.0       correct
NHWC [1,2,31,64] padded=[1,2,32,64] -> canonical [1,1,62,64] / [1,1,64,64] | mean(ones) = 0.984375  wrong
```

The first is the genuine suffix-pad model that #54410 deliberately keeps supported. Since the reshape leaves the reduce with the same `logical_shape` / `padded_shape` pair in both cases, no predicate over those shapes can separate them. The reduce-side `TT_FATAL` is doing everything it can with the information it has.

## Suggested fix

Gate the fast path in `pool2d()` before the canonical reshape, on the spatial padding actually being a suffix of the folded axis:

- no spatial pad at all (`padded[1] == logical[1] && padded[2] == logical[2]`), or
- `logical[1] == 1`, so the only padded spatial dim is the last one and the pad genuinely trails.

Otherwise either strip the padding first or fall through to the sliding-window path. Note the fast path already de-pads TILE inputs via its `to_layout(..., ROW_MAJOR)` conversion — the ROW_MAJOR branch is the one that carries the misleading `padded_shape` through.

Alternatively, reconsider whether the fast path should build a tensor whose declared spec does not describe its data at all: the canonical view is malformed for any consumer, not just reduce.
